mempool/clist_mempool.go

```diff
@@ -649,29 +610,139 @@ func (mem *CListMempool) Update(
 	return nil
 }
 
+// recheckTxs sends all transactions in the mempool to the app for re-validation. When the function
+// returns, all recheck responses from the app have been processed.
 func (mem *CListMempool) recheckTxs() {
-	if mem.Size() == 0 {
-		panic("recheckTxs is called, but the mempool is empty")
+	if mem.Size() <= 0 {
+		return
 	}
+	mem.logger.Debug("recheck txs", "height", mem.height.Load(), "num-txs", mem.Size())
 
-	mem.recheckCursor = mem.txs.Front()
-	mem.recheckEnd = mem.txs.Back()
+	mem.recheck.init(mem.txs.Front(), mem.txs.Back())
 
-	// Push txs to proxyAppConn
-	// NOTE: globalCb may be called concurrently.
+	// NOTE: globalCb may be called concurrently, but CheckTx cannot be executed concurrently
+	// because this function has the lock (via Update and Lock).
 	for e := mem.txs.Front(); e != nil; e = e.Next() {
-		memTx := e.Value.(*mempoolTx)
+		tx := e.Value.(*mempoolTx).tx
+		mem.recheck.numPendingTxs.Add(1)
+
+		// Send a CheckTx request to the app. If we're using a sync client, the resCbRecheck
+		// callback will be called right after receiving the response.
 		_, err := mem.proxyAppConn.CheckTxAsync(context.TODO(), &abci.CheckTxRequest{
-			Tx:   memTx.tx,
+			Tx:   tx,
 			Type: abci.CHECK_TX_TYPE_RECHECK,
 		})
 		if err != nil {
-			mem.logger.Error("recheckTx", "err", err)
-			return
+			mem.logger.Error("failed to send reCheckTx request", "err", err, "tx", tx.Hash())
+			mem.recheck.tryFinish() // in case the tx is the last to recheck and there's no response for it.
 		}
 	}
 
-	// In <v0.37 we would call FlushAsync at the end of recheckTx forcing the buffer to flush
-	// all pending messages to the app. There doesn't seem to be any need here as the buffer
-	// will get flushed regularly or when filled.
+	// Flush any pending asynchronous recheck requests to process.
+	mem.proxyAppConn.Flush(context.TODO())
+
+	// Give some time to finish processing the responses; then finish rechecking, even if not all
+	// txs were rechecked.
+	select {
+	case <-time.After(timeoutRecheck):
```

Comment:
Why? We are now imposing the the application a maximum delay for performing operations?

Reply:
For async ABCI clients, where CheckTx responses may arrive late or never, we have two options. 
- One is to block waiting for CheckTx responses. In particular during rechecking, which happens inside Update, where the updateMtx lock is held. If the response never arrives, Update will block forever.
- The second option is to put a timeout, which is what I did here. It's not critical if some response arrive late and is not processed. IMO this option is better than blocking.